pkg/reconciler/layer3.go

```diff
@@ -0,0 +1,214 @@
+
+	changed, err := r.frrManager.Configure(frr.FRRConfiguration{
+		VRFs: vrfConfigs,
+		ASN:  r.config.ServerASN,
+	})
+	if err != nil {
+		r.Logger.Error(err, "error updating FRR configuration")
+		return err
+	}
+
+	if changed || r.dirtyFRRConfig {
+		r.Logger.Info("trying to reload FRR config because it changed")
+		err = r.frrManager.ReloadFRR()
+		if err != nil {
+			r.dirtyFRRConfig = true
+			r.Logger.Error(err, "error reloading FRR systemd unit")
+			return err
+		}
+		r.dirtyFRRConfig = false
+	}
+
+	// Make sure that all created netlink VRFs are up after FRR reload
+	time.Sleep(2 * time.Second)
+	for _, info := range created {
+		if err := r.netlinkManager.UpL3(info); err != nil {
+			r.Logger.Error(err, "error setting L3 to state UP")
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *reconcile) reconcileL3Netlink(vrfConfigs []frr.VRFConfiguration) ([]nl.VRFInformation, error) {
+	existing, err := r.netlinkManager.ListL3()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check for VRFs that are configured on the host but no longer in Kubernetes
+	delete := []nl.VRFInformation{}
+	for _, cfg := range existing {
+		stillExists := false
+		for _, vrf := range vrfConfigs {
+			if vrf.Name == cfg.Name && vrf.VNI == cfg.VNI {
+				stillExists = true
+			}
+		}
+		if !stillExists {
+			delete = append(delete, cfg)
+		}
+	}
+
+	// Check for VRFs that are in Kubernetes but not yet configured on the host
+	create := []nl.VRFInformation{}
+	for _, vrf := range vrfConfigs {
+		// Skip VRF with VNI SKIP_VRF_TEMPLATE_VNI
+		if vrf.VNI == config.SKIP_VRF_TEMPLATE_VNI {
+			continue
+		}
+		alreadyExists := false
+		for _, cfg := range existing {
+			if vrf.Name == cfg.Name && vrf.VNI == cfg.VNI {
+				alreadyExists = true
+			}
+		}
+		if !alreadyExists {
+			create = append(create, nl.VRFInformation{
+				Name: vrf.Name,
+				VNI:  vrf.VNI,
+			})
+		}
+	}
+
+	// Delete / Cleanup VRFs
+	for _, info := range delete {
+		r.Logger.Info("Deleting VRF because it is no longer configured in Kubernetes", "vrf", info.Name, "vni", info.VNI)
+		errs := r.netlinkManager.CleanupL3(info.Name)
+		for _, err := range errs {
+			r.Logger.Error(err, "Error deleting VRF", "vrf", info.Name, "vni", strconv.Itoa(info.VNI))
+		}
+	}
+	// Create VRFs
+	for _, info := range create {
+		r.Logger.Info("Creating VRF to match Kubernetes", "vrf", info.Name, "vni", info.VNI)
+		err := r.netlinkManager.CreateL3(info)
+		if err != nil {
+			return nil, fmt.Errorf("error creating VRF %s, VNI %d: %w", info.Name, info.VNI, err)
+		}
+	}
+
+	return create, nil
+}
+
+func handlePrefixItemList(input []networkv1alpha1.VrfRouteConfigurationPrefixItem, seq int) frr.PrefixList {
+	prefixList := frr.PrefixList{
+		Seq: seq + 1,
+	}
+	for i, item := range input {
+		prefixList.Items = append(prefixList.Items, copyPrefixItemToFRRItem(i, item))
+	}
+	return prefixList
+}
+
+func copyPrefixItemToFRRItem(i int, item networkv1alpha1.VrfRouteConfigurationPrefixItem) frr.PrefixedRouteItem {
```

Comment:
would call it `n` instead of `i`, since `i` is often used in loops, and `n` for arbitrary numbers or counts.